DSharpPlus.CommandsNext/Attributes/RequireRolesAttribute.cs

```diff
@@ -54,51 +60,130 @@ public RequireRolesAttribute(RoleCheckMode checkMode, params string[] roleNames)
+                var intersectIdCount = intersectIds.Count();
+
+                return this.CheckMode switch
+                {
+                    RoleCheckMode.All => Task.FromResult(this.RoleNames.Count == intersectIdCount),
+                    RoleCheckMode.SpecifiedOnly => Task.FromResult(roleIdCount == intersectIdCount),
+                    RoleCheckMode.None => Task.FromResult(intersectIdCount == 0),
+                    _ => Task.FromResult(intersectIdCount > 0),
+                };
+            }
+            else
+            {
+                var roleIds = ctx.Member.RoleIds
+                    .Concat(ctx.Member.Roles.Select(xm => xm.Id));
+
+                var roleNames = ctx.Member.Roles.Select(xm => xm.Name)
+                    .Concat(ctx.Member.Roles.Select(xm => xm.Name));
+
+                var intersectIds = roleIds.Intersect(this.RoleIds);
+                var intersectNames = roleNames.Intersect(this.RoleNames);
+
+                return this.CheckMode switch
+                {
+                    RoleCheckMode.All => Task.FromResult(this.RoleNames.Count == intersectNames.Count()
+                        && this.RoleIds.Count == intersectIds.Count()),
+                    RoleCheckMode.SpecifiedOnly => Task.FromResult(roleIds.Count() == intersectIds.Count()
+                        && roleNames.Count() == intersectNames.Count()),
+                    RoleCheckMode.None => Task.FromResult(!intersectIds.Any() && !intersectNames.Any()),
+                    _ => Task.FromResult(intersectIds.Any() || intersectNames.Any())
+                };
+            }
         }
     }
 
     /// <summary>
-    /// Specifies how does <see cref="RequireRolesAttribute"/> check for roles.
+    /// Specifies how <see cref="RequireRolesAttribute"/> checks for roles.
     /// </summary>
+    [Flags]
     public enum RoleCheckMode
     {
         /// <summary>
-        /// Member is required to have any of the specified roles.
+        /// Member is required to have none of the specified roles.
         /// </summary>
-        Any,
+        None = 0,
+
 
         /// <summary>
         /// Member is required to have all of the specified roles.
         /// </summary>
-        All,
+        All = 1,
+
+        /// <summary>
+        /// Member is required to have any of the specified roles.
+        /// </summary>
+        Any = 2,
 
         /// <summary>
         /// Member is required to have exactly the same roles as specified; no extra roles may be present.
         /// </summary>
-        SpecifiedOnly,
+        SpecifiedOnly = 4,
 
         /// <summary>
-        /// Member is required to have none of the specified roles.
+        /// Instructs the check to evaluate for matching role names.
+        /// </summary>
+        MatchNames = 8,
+
+        /// <summary>
+        /// Instructs the check to evaluate for matching role IDs.
         /// </summary>
-        None
+        MatchIds = 16
```

Comment:
I think its bad idea to change `RoleCheckMode`, since its idea is **how** to match roles, and your proposal is to make it also decide **by what criteria** to match roles. I think its better idea to create new enum with `MatchIds` and `MatchNames` fields, and accept it from constructors. Also, adding flags to `RoleCheckMode` allows users to make `RoleCheckMode.All | RoleCheckMode.Any | RoleCheckMode.None` which also isn't a good thing.

Reply:
users could do that before (somewhat) by casting an integer; i do not think this is a problem.

Reply:
Yeah, you are right.
But still there is semantic difference between what enum did before and what it also does now. I'm not super strong against it, just giving my opinion, it up to you or library owners to decide if it's valid or not.

Reply:
there is a difference, and i'm not too happy with it, but i dont think the difference warrants a separate enum. i'm fairly certain behaviour did not change, which was my main concern with this.